include/cantera/base/Delegator.h

```diff
@@ -0,0 +1,282 @@
+//! @file Delegator.h
+
+// This file is part of Cantera. See License.txt in the top-level directory or
+// at https://cantera.org/license.txt for license and copyright information.
+
+#ifndef CT_DELEGATOR_H
+#define CT_DELEGATOR_H
+
+#include "cantera/base/global.h"
+#include "cantera/base/ctexceptions.h"
+#include <array>
+
+namespace Cantera
+{
+
+//! Delegate member functions of a C++ class to externally-specified functions
+/*
+ * This base class provides functions for setting delegates for the member
+ * functions of a C++ class at runtime. The purpose of this capability is to
+ * allow the class to be extended using functions defined in any programming
+ * language that provides a C API for calling functions in that language.
+ *
+ * Delegates are specified as std::function objects that are responsible for
+ * encapsulating the data specific to the target language and calling the
+ * appropriate function in the target language. The std::function has a
+ * modified function signature compared to the method that it is replacing or
+ * augmenting:
+ * - Methods with no return value and scalar arguments are treated the same
+ * - Methods with a return value have that value as the first reference argument
+ *   of their delegate function, and return an int. The delegate should return
+ *   zero if it does not set the arguments value, and a non-zero value if it
+ *   does.
+ * - Methods with pointers to arrays as arguments have an additional argument
+ *   introduced to indicate the length of each array argument. This argument
+ *   occurs either first, or after the return value reference, and is a
+ *   std::array<size_t, N> where N is the number of array arguments.
+ *
+ * Delegated methods can be specified to either "replace" the original class's
+ * method, or to run "before" or "after" the original method, using the `when`
+ * parameter of the `setDelegate` method.
+ */
+class Delegator
+{
+public:
+    //! A method overridden in derived classes to set delegates for member
+    //! functions with the signature `void()`.
+    virtual void setDelegate(const std::string& name,
+                             const std::function<void()>& func,
+                             const std::string& when)
+    {
+        throw NotImplementedError("Delegator::setDelegate for void(double)");
```

Comment:
It would be nice to have a unit test covering all the exceptions here to improve code coverage. As an aside, I added the ability to add arguments to `NotImplementedError` some time ago (similar to `CanteraError`), which allows for tailoring of the error messages (e.g. if this is ever called from a derived class, the class name can be provided).